### ISSUE TITLE
KMeans DTW: Inertia increases with more clusters

### ISSUE THREAD

**Issue by WhiteLin3s** [tslearn-team/tslearn#306]
Hey there,

I've got the following problem: I'm trying to identify the optimal number of clusters for my data. Therefore I run the KMeans (DTW) algorithm with different values for "n_clusters", while not changing any other parameter or my data.

I focus on the inertia value and use the "elbow method" to identify the optimal number of clusters. I would expect inertia to decrease every time, when I increase "n_clusters" ceteris paribus. But at some steps, inertia seems to increase even though I use more clusters. I already tried playing around with the paramters, like setting a random seed, varying the initialization of the clusters etc. But I can't manage to get a strictly decreasing inertia value.

Is there an explanation for this behaviour? Does it have sth. to do with the DTW method and Barycenter Clustering? I'm new to time series clustering and would appreciate any help, since I use this package for my ongoing thesis. 

Thanks in advance & Greetings!

**Comment by rtavenar** [comment]
Hi @WhiteLin3s and thanks for your question,

One possible explanation is related to initialization: the methods involved are known to converge to local minima, hence it is possible that this causes the non-decreasing behaviour you observe.
To alleviate this, you could probably define the initial barycenters for `n_clusters=k+` as the final ones from `n_clusters=k` plus one randomly positionned centroid: in this case, you should always observe that a larger number of cluster decreases the kmeans cost.

Best regards,
Romain

**Comment by WhiteLin3s** [comment]
Hey @rtavenar,

thank you very much for the clear explanation and suggestion how to fix the problem. Sounds like this should work, will definitely try it out later!

Best regards

**Comment by tatianameleshko** [comment]
Having the same issue, can you please provide code example how to initialize  barycenters for n_clusters=k+ as the final ones from n_clusters=k plus one randomly positionned centroid? Should we use "init" parameter with "cluster_centers_ " attribute? Thank you

**Comment by rtavenar** [comment]
I would do something like this:

```python
from tslearn.clustering import TimeSeriesKMeans
from tslearn.datasets import CachedDatasets
from tslearn.utils import to_time_series_dataset
import numpy

X = CachedDatasets().load_dataset("Trace")[0]

models = []
km = TimeSeriesKMeans(n_clusters=2, metric="dtw", max_iter=10)
km.fit(X)
models.append(km)

for k in range(3, 6):
    barycenters_init = [centroid for centroid in models[-1].cluster_centers_]
    idx_new_center = numpy.random.choice(X.shape[0])
    barycenters_init.append(X[idx_new_center])
    barycenters_init = to_time_series_dataset(barycenters_init)
    km = TimeSeriesKMeans(n_clusters=k, metric="dtw", max_iter=10, init=barycenters_init)
    km.fit(X)
    models.append(km)

for km in models:
    print(km.inertia_)
```

**Comment by tatianameleshko** [comment]
Thank you! It works with "dtw", for "softdtw" - still seeing increasing trend, but its fine for me -  just switched to "dtw".

**Comment by zhouchengcom** [comment]
Is there a code example for identify the optimal number of clusters ？thinks

**Comment by GillesVandewiele** [comment]
You can use this but replace the sklearn KMeans & silhouette score by those of tslearn.

https://scikit-learn.org/stable/auto_examples/cluster/plot_kmeans_silhouette_analysis.html#sphx-glr-auto-examples-cluster-plot-kmeans-silhouette-analysis-py

Basically: try different k's and measure the silhouette_score. Take the best one.

**Comment by jbbqqf** [comment]
Hi! Triaging older issues — I think the premise here doesn't quite match how DTW k-means works.

For Euclidean k-means, inertia is monotonically non-increasing in `n_clusters` because the objective is convex per cluster (means minimise sum of squares globally). With DTW + DBA k-means (`TimeSeriesKMeans(metric="dtw")`), the per-cluster objective is *non-convex*: DBA optimises by alternating warping-path computation and barycenter update, which only converges to a local minimum. So with a different `n_clusters`, you can:

* land in a different local minimum where the sum of within-cluster DTW distances happens to be larger,
* hit early stopping (`max_iter_barycenter`, `tol`) with a worse barycenter,
* see numerical warping-path tie-breaking land differently.

This means inertia is *not* guaranteed monotonic and the elbow method is best-effort with DTW. Common workarounds in the existing tslearn discussion: pin `random_state`, increase `n_init` and `max_iter_barycenter`, or use the silhouette score with `metric="dtw"` instead of inertia.

If you've got a reproducer where inertia is *strongly* increasing under a fixed seed (not just within numerical noise), I'd be happy to look closer — but otherwise this is expected DBA-k-means behaviour and might be a candidate to close.

(Disclosure: I drafted this comment with help from Claude Code while triaging stale issues; the references above were verified manually against `tslearn/clustering/kmeans.py` and `tslearn/barycenters/dba.py`.)

**Comment by charavelg** [comment]
Closing as answered
